src/audio/dai-zephyr.c

```diff
@@ -1359,37 +1355,41 @@ int dai_zephyr_multi_endpoint_copy(struct dai_data **dd, struct comp_dev *dev,
 		case -EPIPE:
 			/* DMA status can return -EPIPE and current status content if xrun occurs */
 			if (direction == SOF_IPC_STREAM_PLAYBACK)
-				comp_dbg(dev, "dai_zephyr_multi_endpoint_copy(): dma_get_status() underrun occurred, endpoint: %d ret = %u",
+				comp_dbg(dev, "dma_get_status() underrun occurred endpoint: %d ret = %u",
 					 i, ret);
 			else
-				comp_dbg(dev, "dai_zephyr_multi_endpoint_copy(): dma_get_status() overrun occurred, enpdoint: %d ret = %u",
+				comp_dbg(dev, "dma_get_status() overrun occurred, enpdoint: %d ret = %u",
 					 i, ret);
 			break;
 		default:
 			return ret;
 		}
 
-		avail_bytes = MIN(avail_bytes, stat.pending_length);
-		free_bytes = MIN(free_bytes, stat.free);
-	}
-
-	/* calculate minimum size to copy */
-	if (direction == SOF_IPC_STREAM_PLAYBACK) {
-		src_frames = audio_stream_get_avail_frames(&multi_endpoint_buffer->stream);
-		sink_frames = free_bytes / frame_bytes;
-	} else {
-		src_frames = avail_bytes / frame_bytes;
-		sink_frames = audio_stream_get_free_frames(&multi_endpoint_buffer->stream);
+		/* Use frames_aligned here to avoid glitches, especially for streams
+		 * with odd number of channels. Update frames for every dai in a loop.
+		 */
+		if (direction == SOF_IPC_STREAM_PLAYBACK) {
+			frames_aligned = audio_stream_avail_frames_aligned(
+			    &multi_endpoint_buffer->stream, &dd[i]->dma_buffer->stream);
+			frames = MIN(frames, frames_aligned);
+		} else {
+			frames_aligned = audio_stream_avail_frames_aligned(
+			    &dd[i]->dma_buffer->stream, &multi_endpoint_buffer->stream);
+			frames = MIN(frames, frames_aligned);
```

Comment:
This duplicates line 1374. Consider moving this behind `if {} else {}`.

Reply:
Good catch. Thanks